x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java

```diff
@@ -290,117 +265,81 @@ static String getPolicyId(SnapshotInfo snapshotInfo) {
-                    historyStore.putAsync(SnapshotHistoryItem.deletionSuccessRecord(Instant.now().toEpochMilli(),
-                            info.snapshotId().getName(), policyId, repo));
-                }, e -> {
-                    failed.incrementAndGet();
-                    try {
-                        final SnapshotHistoryItem result = SnapshotHistoryItem.deletionFailureRecord(Instant.now().toEpochMilli(),
-                            info.snapshotId().getName(), policyId, repo, e);
-                        historyStore.putAsync(result);
-                    } catch (IOException ex) {
-                        // This shouldn't happen unless there's an issue with serializing the original exception
-                        logger.error(new ParameterizedMessage(
-                            "failed to record snapshot deletion failure for snapshot lifecycle policy [{}]",
-                            policyId), ex);
-                    }
-                }));
-                // Check whether we have exceeded the maximum time allowed to spend deleting
-                // snapshots, if we have, short-circuit the rest of the deletions
-                long finishTime = nowNanoSupplier.getAsLong();
-                TimeValue deletionTime = TimeValue.timeValueNanos(finishTime - deleteStartTime);
-                logger.debug("elapsed time for deletion of [{}] snapshot: {}", info.snapshotId(), deletionTime);
-                TimeValue totalDeletionTime = TimeValue.timeValueNanos(finishTime - startTime);
-                if (totalDeletionTime.compareTo(maximumTime) > 0) {
-                    logger.info("maximum snapshot retention deletion time reached, time spent: [{}]," +
-                            " maximum allowed time: [{}], deleted [{}] out of [{}] snapshots scheduled for deletion, failed to delete [{}]",
-                        totalDeletionTime, maximumTime, deleted, count, failed);
-                    slmStats.deletionTime(totalDeletionTime);
-                    slmStats.retentionTimedOut();
-                    return;
-                }
+            deleteSnapshots(slmStats, deleted, failed, repo, snapshots, allDeletesListener);
+        }
+    }
+
+    /**
+     * Set of all currently deleting {@link SnapshotId} used to prevent starting multiple deletes for the same snapshot.
+     */
+    private final Set<SnapshotId> runningDeletions = Collections.synchronizedSet(new HashSet<>());
+
+    private void deleteSnapshots(SnapshotLifecycleStats slmStats, AtomicInteger deleted, AtomicInteger failed, String repo,
+                                 List<SnapshotInfo> snapshots, ActionListener<Void> listener) {
+
+        final ActionListener<Void> allDeletesListener =
+                new GroupedActionListener<>(ActionListener.map(listener, v -> null), snapshots.size());
+        for (SnapshotInfo info : snapshots) {
+            if (runningDeletions.add(info.snapshotId()) == false) {
```

Comment:
I think we leak memory (it would be very rare) if the next two lines (after the `if`) were to throw an exception.

For example, if `getPolicyId(info)` were to throw an exception (possible since it has an `orElseThrow(...)` clause) we'd never remove the id from the deletions, and we'd never be able to remove the snapshot because we'd always think we were currently removing it.

Not sure what the best way to handle it is, but maybe we can just wrap the following lines to remove it otherwise? I still feel a little uneasy about deleteSnapshot potentially not calling the listener (for example, if its implementation were changed and it threw an exception rather than calling `listener.onFailure(...)`)

Reply:
++ to all of the above. Sorry again for the noise here :( => I think I covered all cases with https://github.com/elastic/elasticsearch/pull/62284/commits/41797fa4102f7c370c204dc30d9d810281e7236c now, let me know what you think :)